### DISCUSSION TITLE
Idea: export variable as input action

### DISCUSSION THREAD

**Discussion by theraot** [godotengine/godot-proposals#7559] [Scripting]
I have a set of scripts where I configure an input action (the name of an action from the input map). While I've been able to use `_get_property_list` to provide a list of options, it is common enough - at least for me - that I believe it could use editor help.

Example of what I've used:

```
@tool
extends Resource

var action_name: StringName = &""

func _get_property_list() -> Array[Dictionary]:
	var list_of_actions:String = ""
	for property in ProjectSettings.get_property_list():
		var property_name:String = property.name
		if not property_name.begins_with("input/"):
			continue

		var action := property_name.trim_prefix("input/")
		if not list_of_actions.is_empty():
			list_of_actions += ","

		list_of_actions += action

	return [
		{
			"name": "action_name",
			"type": TYPE_STRING_NAME,
			"hint": PROPERTY_HINT_ENUM_SUGGESTION,
			"hint_string": list_of_actions,
			"usage": PROPERTY_USAGE_SCRIPT_VARIABLE | PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_STORAGE
		}
	]
```

*2025-04-01: Edited with fixed code (avoid using `InputMap.load_from_project_settings`).*

So, I would want to write:

```
@export_action var action_name: StringName = &""
```

And while we are at it, I would rather have the custom actions at the top, currently I get the build-in ones at the top.

**Comment by Calinou** [comment]
This would also be useful for TouchScreenButton's `action` property, although it should allow entering custom names to support actions that are programmatically created at runtime.

**Comment by baznzab** [comment]
I've spent some time implementing this feature via custom plugin and have some thoughts:
1. Using `InputMap.load_from_project_settings()` inside editor script is insecure because it removes all editor related bindings and breaks its behavior. It's better to get actions directly from EditorSettings, similar to how it's done in [`InputMap::load_from_project_settings()`](https://github.com/godotengine/godot/blob/7233bc6736e6a9e12758a52402660f9a3a892a78/core/input/input_map.cpp#L259).
2. Using enum-like gui for such properties isn't the best. QuickLoad-like gui would suit much better here I think. [`EditorQuickOpen`](https://github.com/godotengine/godot/blob/654132cb9c6ff1f27be0bf325e348e74b3e49fba/editor/editor_quick_open.h#L39) class could be reused. I tried to mimic it in my plugin, you can see it on the video attachment, but it lacks some qol features like search for example. https://github.com/godotengine/godot-proposals/assets/67281067/93ccf5ba-0426-4508-b78c-c967e4dfbce3
3. Although it is possible to achieve such behavior with custom plugin it isn't the best solution cause it requires a wrapper around input action's `StringName`.

Here is my plugin implemented in gdscript.
https://github.com/baznzab/input_action_reference
Posting it just for the reference and to represent what i mentioned above.

**Comment by AThousandShips** [comment]
This will be solved by:
* https://github.com/godotengine/godot/pull/96611

**Comment by theraot** [comment]
With the pull request mentioned above merged, I'll go ahead and close this.
